pkg/bpf/collection.go

```diff
@@ -41,6 +45,110 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
 	return spec, nil
 }
 
+func removeUnreachableTailcalls(spec *ebpf.CollectionSpec) error {
+	type TailCall struct {
+		referenced bool
+		visited    bool
+		spec       *ebpf.ProgramSpec
+	}
+
+	entrypoints := make([]*ebpf.ProgramSpec, 0)
+	tailcalls := make(map[uint32]*TailCall)
+
+	for _, prog := range spec.Programs {
+		if prog.SectionName == "tc" || prog.SectionName == "xdp" {
```

Comment:
Why not check `prog.Type` instead? This is brittle because we also allow these to be prefixes like `tc/foobar`.

Similarly, only programs where `prog.Type == Unspecified` can be tail calls due to our weird section naming. Maybe it's good to not rely on that though.

Reply:
@lmb The kernel doesn't accept Unspecified progs. These are fixed up in `classifyProgramTypes`.

@dylandreimerink Could we remove some of this logic by running tail call pruning after `iproute2Compat` and `classifyProgramTypes`? After those functions, `cilium_calls` will have its `MapSpec.Contents` correctly populated, and I think we can limit the pruning algorithm to progs appearing in cilium_calls only. For example, entrypoint progs will never appear in cilium_calls.

Reply:
Why not check prog.Type instead? Not sure these are set since the `classifyProgramTypes` program that explicitly sets these runs after this one. (I think it might work now even without the classifyProgramTypes)

In any case, I suggest we remove this specific logic, and let the if not a `x/y` section treat as entrypoint logic catch these, that should work regardless of whatever changes may happen in the future.